tests/integration/rdb.tcl

```diff
@@ -416,4 +416,239 @@ start_server {} {
+        after 100
+
+        # change configuration and reload - result should be dict-encoded key
+        r config set hash-max-listpack-entries 0
+        r debug reload
+
+        assert_equal [lsort [r hgetall key]] "1 2 3 4 a b c d"
+        assert_match "*encoding:hashtable*" [r debug object key]
+    }
+}
+
+set server_path [tmpdir "server.dict-to-listpack-test"]
+
+test "save dict, load listpack" {
+    start_server [list overrides [list "dir" $server_path  enable-debug-command yes]] {
+        r config set hash-max-listpack-entries 0
+
+        r FLUSHALL
+
+        r HMSET key a 1 b 2 c 3 d 4
+        assert_match "*encoding:hashtable*" [r debug object key]
+        r HPEXPIRE key 100 1 d
+
+        # sleep 100 ms to make sure all fields will expire after restart
+        after 100
+
+        # change configuration and reload - result should be LP-encoded key
+        r config set hash-max-listpack-entries 512
+        r debug reload
+
+        assert_equal [lsort [r hgetall key]] "1 2 3 a b c"
+        assert_match "*encoding:listpack*" [r debug object key]
+    }
+}
+
+set server_path [tmpdir "server.active-expiry-after-load"]
+
+# verifies a field is correctly expired by active expiry AFTER loading from RDB
+foreach {type lp_entries} {listpack 512 dict 0} {
+    start_server [list overrides [list "dir" $server_path enable-debug-command yes]] {
+        test "active field expiry after load, ($type)" {
+            r config set hash-max-listpack-entries $lp_entries
+
+            r FLUSHALL
+
+            r HMSET key a 1 b 2 c 3 d 4 e 5 f 6
+            r HEXPIREAT key 2524600800 2 a b
+            r HPEXPIRE key 200 2 c d
+
+            r save
+            r debug reload
+
+            # sleep 1 sec to make sure 'c' and 'd' will active-expire
+            after 1000
+
+            assert_equal [s rdb_last_load_keys_loaded] 1
+            assert_equal [s rdb_last_load_hash_fields_expired] 0
+            assert_equal [s expired_hash_fields] 2
+
+            # hgetall might lazy expire fields, so it's only called after the stat asserts
+            assert_equal [lsort [r hgetall key]] "1 2 5 6 a b e f"
+            assert_equal [r hexpiretime key 6 a b c d e f] {2524600800 2524600800 -2 -2 -1 -1}
+        }
+    }
+}
+
+set server_path [tmpdir "server.lazy-expiry-after-load"]
+
+foreach {type lp_entries} {listpack 512 dict 0} {
+    start_server [list overrides [list "dir" $server_path enable-debug-command yes]] {
+        test "lazy field expiry after load, ($type)" {
+            r config set hash-max-listpack-entries $lp_entries
+            r debug set-active-expire 0
+
+            r FLUSHALL
+
+            r HMSET key a 1 b 2 c 3 d 4 e 5 f 6
+            r HEXPIREAT key 2524600800 2 a b
+            r HPEXPIRE key 200 2 c d
+
+            r save
+            r debug reload
+
+            # sleep 1 sec to make sure 'c' and 'd' will lazy-expire when calling hgetall
+            after 1000
```

Comment:
Please trigger a daily build for your branch. Tests that rely on timing might fail on slow environments. 
Better approach is using `wait_for_condition`.

Reply:
How do I trigger a daily build?

Reply:
sorry, you should be able do it on your fork here: https://github.com/ronen-kalish/redis/actions/workflows/daily.yml
Daily -> run workflow -> Run workflow.  (write something dummy to `jobs to skip`, `tests to skip` sections.)

For your pr, I triggered one on my fork: https://github.com/tezc/redis/actions/runs/9098441743
It takes a few hours, fyi.